Initial commit, oauth2 access token parser

baabel opened the pull request:
Added OAUTH2 access token parser which is parsed from http authorization header, query string or body.

baabel committed:
Initial commit, oauth2 access token parser

micahr commented:
Thank you for the PR, love the tests.

Any objections @yunong or @DonutEspresso?

DonutEspresso commented:
@baabel, this is great! For context, it I believe this PR implements [https://tools.ietf.org/html/rfc6750](RFC6750).

According to RFC it looks like we are missing a few validation/verification checks:
- "Clients MUST NOT use more than one method to transmit the token in each request." (Section 2 - you should probably return a 400). 
- "The HTTP request entity-header includes the "Content-Type" header field set to "application/x-www-form-urlencoded"" (Section 2.2 - don't set the token if this isn't true when inspecting req.body)
- "Clients using the URI Query Parameter method SHOULD also send a Cache-Control header containing the "no-store" option.  Server success (2XX status) responses to these requests SHOULD contain a Cache-Control header with the "private" option." (Section 2.3)

Regarding the last, given that using query params is not very secure, and it's not required to implement according to the RFC, I'd be ok if we chopped it entirely. If we can get these extra checks added we should be good to go. Thanks for the contrib!

baabel committed:
Updated to conform to The OAuth 2.0 Authorization Framework: Bearer Token Usage

Specifically updated to fix the following:

According to RFC we are missing a few validation/verification checks:

"Clients MUST NOT use more than one method to transmit the token in each request." (Section 2 - you should probably return a 400).

"The HTTP request entity-header includes the "Content-Type" header field set to "application/x-www-form-urlencoded"" (Section 2.2 - don't set the token if this isn't true when inspecting req.body)

REMOVED URI QUERY STRING PARSING DUE TO SECURITY CONCERNS
"Clients using the URI Query Parameter method SHOULD also send a Cache-Control header containing the "no-store" option. Server success (2XX status) responses to these requests SHOULD contain a Cache-Control header with the "private" option." (Section 2.3)

baabel committed:
Fixed defects around parsing token in body & header

baabel commented:
Hi, I updated the oauth2 token parser per your suggestions.  Ran into an interesting case, if I set the content-type to application/json and POST with an access_token in the body, the core http component generates an authorization header with a bearer token from the body.  In that case the token is sent both as BODY content and HEADER so would fail the 2 method test.

baabel commented:
@DonutEspresso - I updated the OAUTH2 parser per your suggestions and it is ready.  Thanks!

DonutEspresso commented on lib/plugins/oauth2TokenParser.js:
```diff
@@ -0,0 +1,96 @@
+                // more than one method to transmit the token in each request
+                // is not allowed - return 400
+                if (tokenFromBody && tokenFromHeader) {
+                    throw new errors.makeErrFromCode(400);
```
We don't need the try/catch here - you can just return next within this if block, since we don't want to try/catch potential issues from qs. We can also add an error message here for some better visibility:

```
if (tokenFromBody && tokenFromHeader) {
    return next(new errors.makeErrFromCode(400, 'multiple tokens disallowed'));
}
```

DonutEspresso commented:
@baabel changes look good. I had one minor comment, once that's fixed up this looks good to go. Thanks for your time!

baabel committed:
Removed try/catch processing qs.  Added context to multiple token not allowed error message.

baabel committed:
Fixed eslint errors.

micahr commented:
LGTM with the latest changes as well.

DonutEspresso commented on lib/plugins/oauth2TokenParser.js:
```diff
@@ -0,0 +1,90 @@
+            req.oauth2.accessToken = tokenFromHeader;
+        }
+
+        var params = querystring.parse(req.body);
```
Just noticed this bit now that I look at it again this morning - what if user has declared bodyParser() middleware prior to oauthtoken parser? This would not work, since `req.body` is an object, right? This is complicated by the fact that if you do use bodyParser, original post body will become req._body, so it's not consistent. 

I think we should just say if you want to parse oauth tokens out of the body, you must use the bodyParser first. That way, this handler is just in charge of putting on the `oauth2` namespace plus any format validation, if such a thing needs to happen in the future. 

Thus, something like:

```
server.use(plugins.bodyParser());
server.use(plugins.oauth2TokenParser());
```

But if you choose not to use bodyParser, we'll still parse it out of the header. Otherwise, as it is currently is, seems like we're doing extra work parsing the body twice, and in some scenarios (if the postbody is a diff format) this handler would probably be completely hosed. I now realize that's probably what @baabel was trying to account for with the original try/catch - sorry for the churn here. :( 

TL,DR;, let's assume req.body is already parsed here, and if it's not, then don't set the token. That way you can choose to use the token parser to get header parsing by default, but if you add a bodyparser in front you'll get token parsing from the the post body too. 

@micahr @baabel does that sound reasonable?

micahr replied:
Great catch, that does sound reasonable to me.

baabel replied:
Thanks for the catch.  I'll update by assuming req.body is an object and by putting a comment about the dependency on bodyParser for parsing body-based tokens.

baabel committed:
Updated to expect the BODY (req.body) to be an object already parsed by bodyParser
For oauth2 tokens parsed from the BODY there is a dependency on the bodyParser plugin.
If parsing only from the authorization header then the bodyParser is not required.

DonutEspresso commented:
@baabel, thanks for making an effort to push this all the way through. Much appreciated. :+1:
